python/ray/data/dataset.py

```diff
@@ -516,42 +504,46 @@ def map_batches(
+            block: Block,
             batch_fn: BatchUDF,
             *fn_args,
             **fn_kwargs,
         ) -> Iterable[Block]:
             DatasetContext._set_current(context)
             output_buffer = BlockOutputBuffer(None, context.target_max_block_size)
-            # Ensure that zero-copy batch views are copied so mutating UDFs don't error.
-            batcher = Batcher(batch_size, ensure_copy=batch_size is not None)
-            for block in blocks:
-                batcher.add(block)
-            batcher.done_adding()
-            while batcher.has_any():
-                batch = batcher.next_batch()
+            block = BlockAccessor.for_block(block)
+            total_rows = block.num_rows()
+            max_batch_size = batch_size
+            if max_batch_size is None:
+                max_batch_size = max(total_rows, 1)
+
+            for start in range(0, total_rows, max_batch_size):
+                # Build a block for each batch.
+                end = min(total_rows, start + max_batch_size)
+                # Make sure to copy if slicing to avoid the Arrow serialization
+                # bug where we include the entire base view on serialization.
+                view = block.slice(start, end, copy=batch_size is not None)
                 # Convert to batch format.
-                batch = BlockAccessor.for_block(batch).to_batch_format(batch_format)
-                # Apply UDF.
-                batch = batch_fn(batch, *fn_args, **fn_kwargs)
+                view = BlockAccessor.for_block(view).to_batch_format(batch_format)
+
+                applied = batch_fn(view, *fn_args, **fn_kwargs)
                 if not (
-                    isinstance(batch, list)
-                    or isinstance(batch, pa.Table)
-                    or isinstance(batch, np.ndarray)
+                    isinstance(applied, list)
+                    or isinstance(applied, pa.Table)
+                    or isinstance(applied, np.ndarray)
                     or (
-                        isinstance(batch, dict)
-                        and all(isinstance(col, np.ndarray) for col in batch.values())
+                        isinstance(applied, dict)
+                        and all(isinstance(col, np.ndarray) for col in applied.values())
                     )
-                    or isinstance(batch, pd.core.frame.DataFrame)
+                    or isinstance(applied, pd.core.frame.DataFrame)
                 ):
                     raise ValueError(
                         "The map batches UDF returned the value "
-                        f"{batch} of type {type(batch)}, "
+                        f"{applied} of type {type(applied)}, "
                         "which is not allowed. "
                         f"The return type must be one of: {BatchType}"
                     )
-                # Add output batch to output buffer.
-                output_buffer.add_batch(batch)
+                output_buffer.add_batch(applied)
```

Comment:
This is manual change after resolving conflict, please double check @clarkzinzow.